poly/schnorr.go

```diff
@@ -0,0 +1,270 @@
+	// same as the "receiver" index in the joint.go code
+	Index int
+
+	// The partial signature itself
+	Part *abstract.Secret
+}
+
+// SchnorrSig represents the final signature of a distribtued threshold schnorr signature
+// which can be verified against a message
+// This struct is not intended to be constructed manually but can be:
+//  - produce by the Schnorr struct
+//  - verified against a Schnorr struct
+type SchnorrSig struct {
+
+	// the signature itself
+	Signature *abstract.Secret
+
+	// the random public polynomial used during the signature generation
+	Random *PubPoly
+}
+
+// NewSchnorr nstantiate a Schnorr  struct . A wrapper around Init
+func NewSchnorr(info PolyInfo, longterm *SharedSecret) *Schnorr {
+	return new(Schnorr).Init(info, longterm)
+}
+
+// Init initialize the Schnorr struct
+func (s *Schnorr) Init(info PolyInfo, longterm *SharedSecret) *Schnorr {
+	s.info = info
+	s.Longterm = longterm
+	return s
+}
+
+// NewRound set the random key for the d.schnorr algo + set the msg to be signed
+// You call this function when you want a new signature to be issued on a specific message
+// The security of the d. schnorr signature protocol is the same as the regular :
+// The random secret MUST BE FRESH for EACH signature / signed message (hence the 'NewRound')
+func (s *Schnorr) NewRound(random *SharedSecret, msg []byte) error {
+	s.random = random
+	s.hash = nil
+	s.partials = nil // erase the previous partil signature from previous round
+	s.partials = make([]*PartialSchnorrSig, s.info.N)
+	hash, err := s.hashMessage(msg, s.random.Pub.SecretCommit())
+	if err != nil {
+		return errors.New(fmt.Sprintf("Unable to hash the message with the given shared secret : %v", err))
+	}
+	s.hash = &hash
+	return nil
+}
+
+// hash returns a hash of the message and the random secret
+// H( m || V )
+// Returns an error if something went wrong with the marshalling
+func (s *Schnorr) hashMessage(msg []byte, v abstract.Point) (abstract.Secret, error) {
+	vb, err := v.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	c := SUITE.Cipher(vb)
+	c.Message(nil, nil, msg)
+	return SUITE.Secret().Pick(c), nil
+}
+
+// Verify if the received structures are good
+// tests the partials shares if there is some
+func (s *Schnorr) verify() error {
+	if s.Longterm.Index != s.random.Index {
+		return errors.New("The index for the longterm shared secret and the random secret differs for this peer.")
+	}
+	nsig := 0
+	for i, _ := range s.partials {
+		if s.partials[i] != nil {
+			nsig += 1
+		}
+	}
+	if nsig < s.info.T {
+		return errors.New(fmt.Sprintf("Received to few Partial Signatures (%d vs %d) to complete a global schnorr signature", len(s.partials), s.info.T))
+	}
+	return nil
+}
+
+// verifyPartialSig will verify if a given partial signature can be checked against the longterm and random secrets
+// of the schnorr structures.
+func (s *Schnorr) verifyPartialSig(ps *PartialSchnorrSig) error {
```

Comment:
SInce it is a private one, it's fine for me "verifyPart" ! but it has to be consistent with the "RevealPartialSIg" discussion below !